src/hotspot/share/oops/method.cpp

```diff
@@ -2059,198 +2060,68 @@ void BreakpointInfo::clear(Method* method) {
-          }
-        }
-        // Only check each block once for frees.  They're very unlikely.
-        // Increment top past the end of the block.
-        b->_top++;
-      }
-      // need to allocate a next block.
-      if (b->_next == nullptr) {
-        b->_next = _last_free = new JNIMethodBlockNode();
-      }
-    }
-    guarantee(false, "Should always allocate a free block");
-    return nullptr;
-  }
-
-  bool contains(Method** m) {
-    if (m == nullptr) return false;
-    for (JNIMethodBlockNode* b = &_head; b != nullptr; b = b->_next) {
-      if (b->_methods <= m && m < b->_methods + b->_number_of_methods) {
-        // This is a bit of extra checking, for two reasons.  One is
-        // that contains() deals with pointers that are passed in by
-        // JNI code, so making sure that the pointer is aligned
-        // correctly is valuable.  The other is that <= and > are
-        // technically not defined on pointers, so the if guard can
-        // pass spuriously; no modern compiler is likely to make that
-        // a problem, though (and if one did, the guard could also
-        // fail spuriously, which would be bad).
-        ptrdiff_t idx = m - b->_methods;
-        if (b->_methods + idx == m) {
-          return true;
-        }
-      }
-    }
-    return false;  // not found
-  }
-
-  // During class unloading the methods are cleared, which is different
-  // than freed.
-  void clear_all_methods() {
-    for (JNIMethodBlockNode* b = &_head; b != nullptr; b = b->_next) {
-      for (int i = 0; i< b->_number_of_methods; i++) {
-        b->_methods[i] = nullptr;
-      }
-    }
-  }
-#ifndef PRODUCT
-  int count_methods() {
-    // count all allocated methods
-    int count = 0;
-    for (JNIMethodBlockNode* b = &_head; b != nullptr; b = b->_next) {
-      for (int i = 0; i< b->_number_of_methods; i++) {
-        if (b->_methods[i] != _free_method) count++;
-      }
-    }
-    return count;
-  }
-#endif // PRODUCT
-};
-
-// Something that can't be mistaken for an address or a markWord
-Method* const JNIMethodBlock::_free_method = (Method*)55;
-
-JNIMethodBlockNode::JNIMethodBlockNode(int num_methods) : _top(0), _next(nullptr) {
-  _number_of_methods = MAX2(num_methods, min_block_size);
-  _methods = NEW_C_HEAP_ARRAY(Method*, _number_of_methods, mtInternal);
-  for (int i = 0; i < _number_of_methods; i++) {
-    _methods[i] = JNIMethodBlock::_free_method;
-  }
-}
-
-void Method::ensure_jmethod_ids(ClassLoaderData* cld, int capacity) {
-  // Have to add jmethod_ids() to class loader data thread-safely.
-  // Also have to add the method to the list safely, which the lock
-  // protects as well.
-  MutexLocker ml(JmethodIdCreation_lock,  Mutex::_no_safepoint_check_flag);
-  if (cld->jmethod_ids() == nullptr) {
-    cld->set_jmethod_ids(new JNIMethodBlock(capacity));
-  } else {
-    cld->jmethod_ids()->ensure_methods(capacity);
-  }
-}
+// jmethodIDs are 64-bit integers that will never run out and are mapped in a table
```

Comment:
Should we have a `guarantee` or `assert` somewhere that the counter never wraps?

Reply:
Thanks!

Reply:
Given it will take 584 years to wrap-around at a generation rate of 1 per nanosecond, I think we can just use an assertion here, as the only way this could fire is if we initialize the counter incorrectly.